src/simulation.jl

```diff
@@ -190,93 +252,143 @@ function compile_env(d::AbstractNamedDecapode, dec_matrices::Vector{Symbol}, con
 
-    push!(defined_ops, op)
-  end
   return defs
 end
 
-function compile_var(alloc_vectors::Vector{AllocVecCall})
-  return quote $(Expr.(alloc_vectors)...) end
+struct AmbiguousNameException <: Exception
+  name::Symbol
+  indices::Vector{Int}
+end
+
+Base.showerror(io::IO, e::AmbiguousNameException) = print(io, "Name \"$(e.name)\" is repeated at indices $(e.indices) and is ambiguous")
+
+struct InvalidDecaTypeException <: Exception
+  name::Symbol
+  type::Symbol
 end
 
-# This is the block of parameter setting inside f
-# TODO: Pass this an extra type parameter that sets the size of the Floats
-get_vars_code(d::AbstractNamedDecapode, vars::Vector{Symbol}) = get_vars_code(d, vars, Float64)
-
-function get_vars_code(d::AbstractNamedDecapode, vars::Vector{Symbol}, ::Type{stateeltype}) where stateeltype
-  stmts = map(vars) do s
-    ssymbl = QuoteNode(s)
-    if all(d[incident(d, s, :name) , :type] .== :Constant)
-    #if only(d[incident(d, s, :name) , :type]) == :Constant
-      :($s = p.$s)
-    elseif all(d[incident(d, s, :name) , :type] .== :Parameter)
-      :($s = (p.$s)(t))
-    elseif all(d[incident(d, s, :name) , :type] .== :Literal)
-      # Literals don't need assignments, because they are literals, but we stored them as Symbols.
-      # #TODO: we should fix that upstream so that we don't need this.
-      :($s = $(parse(stateeltype, String(s))))
-    else
-      # TODO: If names are not unique, then the type is assumed to be a
-      # form for all of the vars sharing a same name.
-      :($s = u.$s)
+Base.showerror(io::IO, e::InvalidDecaTypeException) = print(io, "Variable \"$(e.name)\" has invalid type \"$(e.type)\"")
+
+"""
+    get_vars_code(d::SummationDecapode, vars::Vector{Symbol}, ::Type{stateeltype}) where stateeltype
+
+This initalizes all input variables according to their Decapodes type.
+"""
+function get_vars_code(d::SummationDecapode, vars::Vector{Symbol}, ::Type{stateeltype}, code_target::GenerationTarget) where stateeltype
+  stmts = quote end
+
+  map(vars) do s
+    # If name is not unique (or not just literals) then error
+    found_names_idxs = incident(d, s, :name)
+
+    # TODO: we should handle the case of same literals better
+    is_singular = length(found_names_idxs) == 1
+    is_all_literals = all(d[found_names_idxs, :type] .== :Literal)
+
+    if isempty(found_names_idxs) || (!is_singular && !is_all_literals)
+      throw(AmbiguousNameException(s, found_names_idxs))
+    end
+
+    s_type = is_all_literals ? :Literal : getgeneric_type(d[only(found_names_idxs), :type])
+
+    # Literals don't need assignments, because they are literals, but we stored them as Symbols.
+    # TODO: we should fix that upstream so that we don't need this.
+    line = @match s_type begin
+      :Literal => :($s = $(parse(stateeltype, String(s))))
+      :Constant => :($s = p.$s)
+      :Parameter => :($s = (p.$s)(t))
+      _ => hook_GVC_get_form(s, s_type, code_target) # ! WARNING: This assumes a form
```

Comment:
Can you use `is_form` to check that? Or is that too difficult to do inside an `@match`?

Reply:
Not sure you can call functions inside of MLStyle.